case_conversion/case_parse.py

```diff
@@ -7,75 +7,199 @@
     unicode = str
 
 
-def parse_case(var, detect_acronyms=True, acronyms=[], preserve_case=False):
-    """
-    Parses a variable into a list of words.
-    Also returns the case type, which can be one of the following:
-        - upper: All words are upper-case.
-        - lower: All words are lower-case.
-        - pascal: All words are title-case or upper-case. Note that the
-                  variable may still have separators.
-        - camel: First word is lower-case, the rest are title-case or
-                 upper-case. Variable may still have separators.
-        - mixed: Any other mixing of word casing. Never occurs if there are
-                 no separators.
-        - unknown: Variable contains no words.
-    Also returns the first separator character, or False if there isn't one.
+UPPER = regex.compile(u'^[\p{Lu}]$')
+SEP = regex.compile(u'^[^\p{Ll}\p{Lu}\p{Nd}]$')
+NOTSEP = regex.compile(u'^[\p{Ll}\p{Lu}\p{Nd}]$')
 
-    TODO: include unicode characters.
-    """
-    # lower = regex.compile('^[\p{Ll}\p{Nd}]$')
-    upper = regex.compile(u'^[\p{Lu}]$')
-    sep = regex.compile(u'^[^\p{Ll}\p{Lu}\p{Nd}]$')
-    notsep = regex.compile(u'^[\p{Ll}\p{Lu}\p{Nd}]$')
 
+def _determine_case(was_upper, words, string):
+    case_type = 'unknown'
+    if was_upper:
+        case_type = 'upper'
+    elif string.islower():
+        case_type = 'lower'
+    elif len(words) > 0:
+        camel_case = words[0].islower()
+        pascal_case = words[0].istitle() or words[0].isupper()
+
+        if camel_case or pascal_case:
+            for word in words[1:]:
+                c = word.istitle() or word.isupper()
+                camel_case &= c
+                pascal_case &= c
+                if not c:
+                    break
+
+        if camel_case:
+            case_type = 'camel'
+        elif pascal_case:
+            case_type = 'pascal'
+        else:
+            case_type = 'mixed'
+
+    return case_type
+
+# Check a run of words represented by the range [s, i]. Should
+# return last index of new word groups.
+
+
+def _advanced_acronym_detection(s, i, words, acronyms):
+    # Combine each letter into single string.
+    acstr = ''.join(words[s:i])
+
+    # List of ranges representing found acronyms.
+    range_list = []
+    # Set of remaining letters.
+    not_range = set(range(len(acstr)))
+
+    # Search for each acronym in acstr.
+    for acronym in acronyms:
+        # TODO: Sanitize acronyms to include only letters.
```

Comment:
Is the idea behind this `TODO` valid?
What about acronymes that do contain letters?
Like `HTML2JADE` or `WW2`?
I do not think we should be as strict here.

Reply:
Agreed. That was a hold over from Davis's code.